### ISSUE TITLE
Disk cache not working properly in libtorrent: 0.16.18.0. Optimize for flash-memory-lifespan

### ISSUE THREAD

**Issue by softwarecreations** [arvidn/libtorrent#1997]
libtorrent: 0.16.18.0
platform/architecture: Raspberry Pi 2B+ armv7l
Installed with latest Deluge: 1.3.10 from Debian Jessie repos.

I recently ran deluged writing to an 8GB ADATA microSD card and it failed after a few hours (will not accept any more writes). So now I'm trying to setup my downloads to be kinder to the flash memory.

I have set the cache size from Deluge to be 5120 * 16 KiB blocks expiring every 3600 seconds.

I've got loads of free memory. I don't have any swap enabled. All log and temp dirs (/tmp /var/tmp /var/log /run /var/cache) etc are mounted as tmpfs. Nothing besides deluged should be writing to disk on the system.
But when I monitor disk writes, I see odd numbers of KiB (that do not fit properly into flash cell sizes) are written to the microSD card every minute.

What I would like to see instead is that deluge should only write flash memory cell sized blocks of let's say 8MiB (to be really conservative) (or at least 1MiB but it can be user settable). And it should only write less than that if the write cache expires.

Taking things a step further, flash memory would have the best lifetime/durability if libtorrent fetches pieces that will fit into an 8MiB sized memory cell. So it's fine to fetch random parts of the torrent, but it should do so in full flash-memory-cell sized blocks. 

```
root@rpi:/sys/block/mmcblk0/queue/iosched# iostat -dk 60
Linux 4.9.24-v7+ (rpi) 	15/05/17 	_armv7l_	(4 CPU)

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           2.44       306.29        51.22    5358207     896016

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.60         0.00        70.33          0       4220

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.62         8.53        46.07        512       2764

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.40         8.53        51.40        512       3084

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.53         8.53        76.13        512       4568

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.48         8.53        11.33        512        680

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.48         8.53        66.33        512       3980

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.43         1.60        40.13         96       2408

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.72         5.00        74.47        300       4468

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.92        17.07        30.33       1024       1820

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.52         7.20        46.40        432       2784

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.65         6.67        58.60        400       3516

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.75         0.00        66.87          0       4012

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.55         0.00        66.40          0       3984

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           2.88        28.47        46.87       1708       2812

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.62         0.67        47.06         40       2824

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.58         8.53        43.87        512       2632

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.72         0.00        54.40          0       3264

```

To confirm if deluged was responsible for doing all the writes I paused all my downloads. This is what happened afterwards.

```
Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.08         0.00         1.40          0         84

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.28         0.00         3.93          0        236

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.00         0.00         0.00          0          0

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.27         0.93         1.93         56        116

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.08         0.00         3.00          0        180

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.17         0.00         0.87          0         52

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
mmcblk0           0.07         0.00         0.33          0         20
```

So the above proves that deluged was writing to disk every minute, and that it was writing blocks much smaller than the cache size. It's hard to say the actual size of data written by deluge, but the above suggested improvements would certainly have a very positive impact on flash lifespan.

I looked into various methods to try improve the flash durability, if you're interested you can check out my serverfault post: https://serverfault.com/questions/850120/aggressive-linux-write-cache-settings-caused-unresponsive-system-and-eventually

I tried to find out if there's anything else on my system writing to disk. With all deluged downloads paused I ran the following command and there were no writes to the microSD card during the few minutes that I monitored it.
`inotifywait -mr @/dev @/sys @/proc @/run @/var/log @/tmp /`
(remember that /var/log and /tmp etc are mounted as tmpfs on my system)

**A very crude and undesirable workaround**
For now I've set deluge to use 12800*16KiB blocks for cache and made the cache expire after 86400 seconds (1 day), in an attempt to prevent unnecessary writes. That means I'm sacrificing 200MiB of RAM which a massive amount of RAM on the RbPi 2 (which has less than 1GB of usable RAM).

If my above idea is implemented then libtorrent would only need 1 or a few multiples of the flash memory block size, which is probably between 1MB or 4MB on most microSD cards or SSD's.
[In 2014 Intel recommended an allocation size of 1MiB](http://www.intel.com/content/www/us/en/solid-state-drives/ssd-partition-alignment-tech-brief.html). So I think this should be the minimum user-selectable size for libtorrent to collect pieces of the torrent. Meaning if the corrent is divided into 512KiB blocks, libtorrent will grab blocks 10 and 11 and then write them together into a single block, instead of grabbing 10, writing that which causes 1 cell write, then grabbing 12 then writing that, causing another cell write, then grabbing 11, causing another cell write etc etc. Ideally we should have 1 cell write for each of the cells occupied by the downloaded torrent.

**Comment by arvidn** [comment]
The disk cache doesn't work the way you think it does. expiring pieces is only relevant for the *read* cache. dirty blocks are always flushed when the piece completes, but also when the cache is running low on space.

There is a feature already that create an affinity for greater-than-piece-size chunks, to be written contiguously to disk. It's not terribly well-supported and really more of an experiment than something to rely on. See the setting for ``allow_partial_disk_writes``

**Comment by Seeker2** [comment]
I download to a ramdrive and then have my BitTorrent client move that to the final destination drive.
It's seamless enough, but requires crazy amounts of ram and/or limited to small downloads at a time.

uTorrent writes entire pieces to the drive in 1 go. While that's not a perfect fit for a SSD, it's far kinder than the 16KB-at-a-time write amplification nightmare that qBitTorrent (and probably Deluge as well) currently is. More details at:   https://qbforums.shiki.hu/index.php/topic,3743.msg19531.html#msg19531

**Comment by arvidn** [comment]
with sufficient cache, libtorrent writes whole pieces at a time. If the disk cache fills up before completing a piece, smaller blocks may be flushed. However, if your piece size is 16 kiB, that will also be the write size.

You may want to make sure you haven't disabled the OS' disk cache as well.

Also, I really think you should use a newer version of libtorrent. If you're going to use a 3 year old version of libtorrent, any potential fix for this problem would still take you 3 years before you pick it up. Not to mention all the fixes that has gone in in the last 3 years.

**Comment by senposage** [comment]
this still occurs even on 1.1.5.0 
platter drives can experience 
extreme thrashing when the internet bandwidth is greater then the disk bandwidth (platter drives are pretty slow at writing lots of small chunks) '

I wonder if some additional buffering might be possible if you have a lot of small pieces buffer them keep them in memory until you can assemble a contiguous block and then write them out
windows caching is unreliable at times,especially if buffer flushing is still enabled

**Comment by arvidn** [comment]
@senposage by deleting the post, are you implying that this is no longer happening in 1.1.x?

**Comment by senposage** [comment]
No it was just badly written I was tired 
the issue still exists in certain cases with small chunk sizes

**Comment by senposage** [comment]
yes when torrents require small writes can overload the windows disk cache and that leads to thrashing its especially a problem on slower 5400RPM platter drives

there's lots of small writes being requested and I am not sure but for small writes I think windows 10 just ignores the cache

**Comment by arvidn** [comment]
@senposage my impression is that most distros still use an old version of libtorrent where ``disable_os_cache`` is still available (but probably deprecated).

the cache in libtorrent *does* coalesce writes, but it never coalesces into larger blocks than a *piece* (so, torrents with small pieces won't get much coalescing).

The other case where coalescing will fail is if there are too many peers being downloaded from in parallel, and the cache size is set too small that it can't fit a full piece from them all. When the cache fills up, any read-cache pieces are evicted to make space, once they are all evicted "dirty" blocks start to be flushed to disk prematurely. But the primary objective of the cache is to maximise the coalescing of writes. (technically it is to minimise the read-back from performing hash-checks, but that behaves very similarly to maximising coalescing).

Perhaps one issue on windows is that, since there's no direct equivalence of ``pwritev()``, I implement that as a series of asynchronous writes of 16 kiB each. Since they're all async (or overlapped in widows terminology) they should reasonably be coalesced by the filesystem/cache-layer. That doesn't explain raspberry-pi issues though. Perhaps the issue there is that the kernels are too old to *have* ``pwritev()`` in the first place, and it ends up being a series of calls to ``pwrite()`` instead.

**Comment by arvidn** [comment]
here's a patch to make the coalesce setting work on windows: https://github.com/arvidn/libtorrent/pull/2738

**Comment by softwarecreations** [comment]
@arvidn Thanks. I'm trying to understand under what conditions the disk cache would be exhausted or ineffective.
So let's say for example that I have 20 peers and I'm downloading a file with 500 pieces, each 8MiB. Let's say libtorrent downloads a fraction of a piece from peer 0, but then that peer goes offline. What will happen to that piece? Will libtorrent delete the partial piece and try download it from the beginning, or will it try to finish downloading that piece from a different peer?
Does libtorrent ever try to download 1 piece from more than 1 peer?
So for example it requests the first half of a piece from one peer, and the 2nd half of a piece from a different peer?
It would help us understand how the caching and downloading strategies work.
It would also be amazing if there can be a debug feature to see visually all the cached pieces downloading, and some cache stats. Like number of pieces in cache, number of complete pieces written, number of incomplete pieces written. So we can better understand how well the cache is working.

Also if we can form some guidelines for how many peers to enable for a given cache size. Some devices are contrained in terms of RAM and use flash storage like a Raspberry Pi. It would be very helpful to have guidelines but also to be able to verify that the number of peers and cache settings are working effectively.

**Comment by arvidn** [comment]
> Let's say libtorrent downloads a fraction of a piece from peer 0, but then that peer goes offline. What will happen to that piece? Will libtorrent delete the partial piece and try download it from the beginning, or will it try to finish downloading that piece from a different peer?

It will not be thrown away. In fact, chances are that some other blocks in that piece was requested and downloaded from other peers as well.
What happens when a peer disconnects is that the blocks requested from it are cleared up to be requested from other peers.

> Does libtorrent prioritize finishing downloading pieces that are almost complete? (when it connects to a new peer)

libtorrent primarily prioritize piece picks based on piece rarity. However, whether there are downloaded blocks in a piece or not is used as a tie-break, where completing pieces that's already been started is prioritized.

> Does libtorrent ever try to download 1 piece from more than 1 peer?

Yes, that is the default. There are a few exceptions. If the download speed from a peer is so high that a whole piece can be completed in 30 seconds (or so), there is an affinity to request whole pieces at a time from it. The other example is if a peer is suspected of sending corrupt data, we request one full (low priority) piece from it, to see if it is (this is called parole mode).

> So for example it requests the first half of a piece from one peer, and the 2nd half of a piece from a different peer?

Pieces are broken down into 16 kiB blocks. Any block in a piece can be requested from any peer. Normally they are picked from left to right, and the number of block requests to keep outstanding to a peer is determined by the download rate we're getting from it. So if you have lots of slow peers, you'll get 2 blocks requested for each peer, all potentially in the same piece.

> Does libtorrent verify the checksum of a piece before it writes it to disk, or after? (it should be before)

Those things do not have a strict dependency. There's an affinity to flush blocks to disk from left to right, and make sure the sha1-hash context is updated as the piece populates. This means most of the time a piece doesn't have to be read-back from disk to check the hash, because the hash was computed incrementally as it was downloaded. So, the cache flush logic has an affinity to flush blocks that have already contributed to the hash computation (to minimize read-back), but ultimately, blocks may need to be flushed to not exceed the cache size limit.

For instance, it would not be reasonable to always keep partial pieces entirely in RAM until they completed download, just to hold off the entire piece from begin flushed before the hash could be checked. For large pieces, that could end up using way more RAM than what's reasonable.

> It would help us understand how the caching and downloading strategies work.
> It would also be amazing if there can be a debug feature to see visually all the cached pieces downloading, and some cache stats. Like number of pieces in cache, number of complete pieces written, number of incomplete pieces written. So we can better understand how well the cache is working.

This is already available. The ``client_test`` that comes with libtorrent can downloading pieces and the state of each block.

> Also if we can form some guidelines for how many peers to enable for a given cache size. Some devices are contrained in terms of RAM and use flash storage like a Raspberry Pi. It would be very helpful to have guidelines but also to be able to verify that the number of peers and cache settings are working effectively.

I don't think guidance can be quite as simple as the number of peers per cache size. If all your peers are really "high quality", and they all have every piece, you can virtually download one piece at a time (not putting a lot of pressure on the cache). But if you have lots of slow peers, that have separate parts of the torrents, they will each maintain a downloading piece, putting maximum pressure on the cache.
